#161966974 Add functionality for user login

EmmanuelBeja opened the pull request:
#### What does this PR do?
This PR delivers the functionality that allows users to login with their accounts on `Author's Haven`.
#### Description of Task to be completed?
- Build out the `Login Form Component`
- Configure the necessary `Actions and Reducers`
- Link the `Login` component with the rest of the `Author's Haven` Routes.
- Connect to backend and consume the `API`#### How should this be manually tested?
1. Execute the following commands in Terminal:
```
git clone https://github.com/andela/ah-jumanji-frontend.git
git checkout ft-login-user-161966974
cd ah-jumanji-frontend
touch .env
```
Add this to `.env`:
```
baseUrl='https://ah-jumanji-staging.herokuapp.com'
loginUrl='/api/users/login'
```
Continue in Terminal
```
npm install
npm start (# or `npm test` to run tests)
``` 

#### Any background context you want to provide?

#### What are the relevant pivotal tracker stories?
[#161966974](https://www.pivotaltracker.com/story/show/161966974)
#### Screenshots (if appropriate)
_Login page_
<img width="1277" alt="screen shot 2019-01-11 at 11 05 46" src="https://user-images.githubusercontent.com/19776238/51020900-f74cf300-1590-11e9-83d1-1eaeb3ec170b.png">

_On Login Fail_
<img width="1277" alt="screen shot 2019-01-16 at 14 14 43" src="https://user-images.githubusercontent.com/19776238/51245588-2f8d7080-1999-11e9-88ad-eeb85fa39b89.png">

_On Login Success_
<img width="1278" alt="screen shot 2019-01-16 at 14 22 47" src="https://user-images.githubusercontent.com/19776238/51245974-454f6580-199a-11e9-944b-d6c5ee38dea5.png">


#### Questions:

#### Checklist:

- [x] My code follows the style guidelines of this project
- [x] At least 2 people have reviewed my PR
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] My PR has one commit.

gwako94 committed:
ft(Login): implement login feature

- Set up initial structure

[Starts #161966974]

gwako94 committed:
ft(Login): implement login feature  …
- Add login form component
- Save data to redux store

[Starts #161966974]

Commit:
ft(Login): implement login feature
- Add login form component
- Save data to redux store
- Display Success and Failure messages

[Starts #161966974]

mashafrancis commented on package.json:
```diff
@@ -88,5 +90,9 @@
     "redux-thunk": "^2.3.0"
+  },
+  "jest": {
+    "snapshotSerializers": ["enzyme-to-json/serializer"],
```
For jest configurations, you can add them to the `jest.config.json` file on the root folder. Reason is not to have a lot of jest config on the `package.json` file.

EmmanuelBeja replied:
done

mashafrancis commented on src/setupTests.js:
```diff
@@ -0,0 +1,5 @@
+
+import Adapter from 'enzyme-adapter-react-16';
+
+configure({ adapter: new Adapter() });
```
There is an `enzyme.config.js` file on the root folder which serves the same purpose for setting up the tests. The same file has been configured at the `jest.config.json`.

EmmanuelBeja replied:
done

gwako94 committed:
ft(Login): implement login feature

- Add login form component
- Save data to redux store
- Display Success and Failure messages
- Add tests

[Starts #161966974]

Commit:
ft(Login): implement login feature
 - Add login form component
 - Save data to redux store
 - Display Success and Failure messages
 - Add tests

[Starts #161966974]

Commit:
ft(Login): implement login feature
 - Add login form component
 - Save data to redux store
 - Display Success and Failure messages
 - Add tests

[Finishes  #161966974]

fabzer0 commented:
Well done

mashafrancis commented on src/apps/Login/components/FormInput.js:
```diff
@@ -0,0 +1,27 @@
+};
+
+FormInput.propTypes = utilsPropTypes;
+
```
Hi. Awesome work on getting the work done. Can I request a minor change that you have this file on the common components folder so that we can have it reusable with our features too since we have the forms with the same input fields etc. It'll help a lot since we won't have to redefine this again on our side.

ghost commented:
> Getting the following `console warnings` when I test this. Seems your `inputForms` have not provided all the required props. It may be necessary to refactor, maybe point to dummy values for what is required that you have not provided/don't need.
> ![image](https://user-images.githubusercontent.com/45358371/51247487-50a49000-199e-11e9-8925-ca4ae4e7475d.png)

Otherwise, works as expected. Good work 👍

EmmanuelBeja commented:
I have resolved that issue @dmithamo1

brayovsky commented on .codeclimate.yml:
```diff
@@ -8,3 +8,5 @@ plugins:
         - .js
         - .jsx
+exclude_patterns:
+- "**/*.test.js"
```
This might not ignore all test files. I'd prefer you remove it so we might have some codeclimate checks apply to test files too and ignore those that might be too extreme like repetition.

brayovsky commented on src/apps/Common/components/WelcomeSection.test.js:
```diff
@@ -0,0 +1,60 @@
+
+    expect(component).toMatchSnapshot();
+  });
+});
```
Can you also add a test to ensure that the `onChange` function is called when `onChange` event occurs?

brayovsky commented on src/apps/Login/actions/loginActions.js:
```diff
@@ -0,0 +1,57 @@
+import * as types from  './actionTypes';
+
+
+const endpoint = process.env.baseUrl + process.env.loginUrl;
```
Add this to a config folder as opposed to an environment variable. A folder preferably under `src/apps/config/config.js` that exports an object with all configurations.
